### ISSUE TITLE
Behavior of jobs without dependencies

### ISSUE THREAD

**Issue by reynir** [mirage/mirage#1426]
In the following code we have a unikernel with a job that does not depend on any devices, but logs a message.

`config.ml`:
```OCaml
open Mirage

let hello = main "Unikernel.Hello" job
let () = register "hello-nodev" [ hello ]
```
`unikernel.ml`:
```OCaml
module Hello = struct
  (* this is imediately evaluated *)
  let start =
    Logs.info (fun f -> f "hello");
    Lwt.return_unit
end
```
This successfully compiles and runs, but on the unix target this does not print anything to the console. My suspicion is the log reporter is not set up before `Unikernel.Hello.start` is evaluated and therefore nothing is printed.

Then my idea was to suspend the computation with a dummy/null/unit device. It was not very clear, and I had to consult the source code, but I found that I could use `Mirage.noop` in `config.ml`:
```OCaml
open Mirage

let hello = main "Unikernel.Hello" (job @-> job)
let () = register "hello-nodev" [ hello $ noop ]
```
And in `unikernel.ml`:
```OCaml
module Hello (_ : sig end) = struct
  let start () =
    Logs.info (fun f -> f "hello");
    Lwt.return_unit
end 
```
I guess for the first code we can't do much about the evaluation order(??) to ensure the log reporter is registered first.

My question then is what is a job? The documentation says a "main task" (but we can have a list of them!?). Should we have a unit typ and unit impl? Can we chain jobs like this?

This was discovered with @PizieDust.

**Comment by palainp** [comment]
Hi @reynir and @PizieDust for finding this :)

I'm not sure if this is a plausible path to track this issue but looking in `caml_program`, calling `camlDune__exe__Unikernel__entry`, it seems that the `start` function is inline in the first version.

I then tried to un-inline it but, as it's no longer a function, the compiler doesn't let me to do that :/.

I'm now at the point where I've added a single unit to the `start` function when no args are provided (see https://github.com/mirage/mirage/compare/main...palainp:mirage:fix-nodev). This is a breaking change as now an unikernel without functorized device will have to take a unit as argument (but I'm not sure this was used before).
@mirage/core, any thoughts on this possible fix?

**Comment by hannesm** [comment]
There's #873 which looks very similar to this issue. I agree with @palainp solution to add a unit argument to `start` functions without any dependencies.

**Comment by reynir** [comment]
Thank you for looking into this! When I was going through mirage-skeleton and playing with mirage with @PizieDust my intuition was that zero device jobs would be a special case with a unit argument. I'm not sure if others share that intuition. On the other hand it can seem inconsistent that it's special cased.

**Comment by hannesm** [comment]
fixed by #1428
